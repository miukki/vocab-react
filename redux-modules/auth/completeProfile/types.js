const prefix = 'auth/completeProfile'

const types = {}
types.set = function(type) {
  this[type] = `${prefix}/${type}`
}

types.set('SHOW')
types.set('HIDE')

types.set('CHANGE_NAME')
types.set('SUBMIT')
types.set('SUBMIT_DONE')
types.set('SHOW_ERROR')
types.set('DISMISS_ERROR')
types.set('RESET_STATE')

export default types
