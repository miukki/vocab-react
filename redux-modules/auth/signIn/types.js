const prefix = 'auth/signIn'

const types = {}
types.set = function(type) {
  this[type] = `${prefix}/${type}`
}

types.set('SHOW')
types.set('HIDE')

types.set('CHANGE_MOBILE')
types.set('CHANGE_PASSWORD')
types.set('SHOW_PASSWORD')
types.set('HIDE_PASSWORD')
types.set('SUBMIT')
types.set('SUBMIT_DONE')
types.set('SHOW_ERROR')
types.set('DISMISS_ERROR')
types.set('GO_FORGOT_PASSWORD')
types.set('RESET_STATE')

export default types
