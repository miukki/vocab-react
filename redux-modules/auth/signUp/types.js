const prefix = 'auth/signUp'

const types = {}
types.set = function(type) {
  this[type] = `${prefix}/${type}`
}

types.set('SHOW')
types.set('HIDE')

types.set('CHANGE_MOBILE')
types.set('SEND_VCODE')
types.set('SEND_VCODE_DONE')
types.set('COUNTDOWN_VCODE')
types.set('CHANGE_VCODE')
types.set('CHANGE_PASSWORD')
types.set('SHOW_PASSWORD')
types.set('HIDE_PASSWORD')
types.set('SUBMIT')
types.set('SUBMIT_DONE')
types.set('SHOW_ERROR')
types.set('DISMISS_ERROR')
types.set('RESET_STATE')

export default types
