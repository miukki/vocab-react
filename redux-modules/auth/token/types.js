const prefix = 'auth/token'

const types = {}
types.set = function(type) {
  this[type] = `${prefix}/${type}`
}

types.set('SET_JWT')
types.set('DELETE_JWT')

export default types
