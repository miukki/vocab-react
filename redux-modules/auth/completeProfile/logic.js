import { createLogic } from 'redux-logic'
import CustomerIamApiSdkJs from 'customer-iam-api-sdk-js'
import { NavigationActions } from 'react-navigation'
import typesSignUp from '../signUp/types'
import types from './types'
import Validator from '../../../lib/Validator'

const getJsState = function(getState) {
  return getState().toJS().completeProfile
}
const getJsStateToken = function(getState) {
  return getState().toJS().token
}

const showCompleteProfileAfterSignUp = createLogic({
  type: typesSignUp.SUBMIT_DONE,
  process({ getState, action }, dispatch, done) {
    dispatch({ type: typesSignUp.HIDE })
    dispatch({ type: types.SHOW })
    done()
  }
})

const validate = createLogic({
  type: [types.SUBMIT],
  validate({ getState, action }, allow, reject, done) {
    const state = getJsState(getState)
    let errorTypes = []
    if (!Validator.isNameCorrect(state.name)) {
      errorTypes.push({ errorType: 'wrongName' })
    }
    if (errorTypes.length) {
      reject({ type: types.SHOW_ERROR, payload: { errorTypes } })
    }
    allow(action)
  }
})

const submitProfile = createLogic({
  type: types.SUBMIT,
  process({ customerIamApiSdkJs, Storage, getState, action }, dispatch, done) {
    //call API
    customerIamApiSdkJs
      .patchUser(getJsState(getState).name, getJsStateToken(getState).jwt)
      .then(res => {
        dispatch({
          type: types.SUBMIT_DONE
        })
        dispatch(NavigationActions.navigate({ routeName: 'Tab' }))
        done()
      })
      .catch(error => {
        if (error === CustomerIamApiSdkJs.ERRORS.NETWORK_ERROR) {
          let errorTypes = []
          errorTypes.push({ errorType: 'networkError' })
          dispatch({
            type: types.SHOW_ERROR,
            payload: { errorTypes }
          })
          return done()
        }

        if (error === CustomerIamApiSdkJs.ERRORS.SERVER_ERROR) {
          let errorTypes = []
          errorTypes.push({ errorType: 'serverError' })
          dispatch({
            type: types.SHOW_ERROR,
            payload: { errorTypes }
          })
          return done()
        }

        if (error === CustomerIamApiSdkJs.ERRORS.USER_PASSWORD_DISMATCH) {
          let errorTypes = []
          errorTypes.push({ errorType: 'userPasswordMismatch' })
          dispatch({
            type: types.SHOW_ERROR,
            payload: { errorTypes }
          })
          return done()
        }

        let errorTypes = []
        errorTypes.push({ errorType: 'otherError' })
        dispatch({
          type: types.SHOW_ERROR,
          payload: { errorTypes }
        })
        done()
      })
  }
})

export default [validate, showCompleteProfileAfterSignUp, submitProfile]
