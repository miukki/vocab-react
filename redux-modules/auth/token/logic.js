// import { NavigationActions } from 'react-navigation'
import { createLogic } from 'redux-logic'
import typesSignUp from '../signUp/types'
import typesSignIn from '../signIn/types'
import typesCompleteProfile from '../completeProfile/types'
import typesForgotPassword from '../forgotPassword/types'
import types from './types'
import { Navigator } from '../../../routers/Root'

const signUpShow = createLogic({
  type: types.DELETE_JWT,
  process({ getState, action }, dispatch, done) {
    dispatch(Navigator.router.getActionForPathAndParams('Login'))
    dispatch({
      type: typesSignUp.RESET_STATE
    })
    dispatch({
      type: typesForgotPassword.RESET_STATE
    })
    dispatch({
      type: typesCompleteProfile.RESET_STATE
    })
    dispatch({
      type: typesSignIn.RESET_STATE
    })
    done()
  }
})

export default [signUpShow]
