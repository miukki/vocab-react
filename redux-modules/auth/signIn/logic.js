import { createLogic } from 'redux-logic'
import CustomerIamApiSdkJs from 'customer-iam-api-sdk-js'
import { NavigationActions } from 'react-navigation'
import * as tokenActions from '../token/actions'
import types from './types'
import Validator from '../../../lib/Validator'

const getJsState = function(getState) {
  return getState().toJS().signIn
}

const validate = createLogic({
  type: [types.SUBMIT],
  validate({ getState, action }, allow, reject, done) {
    const state = getJsState(getState)
    let errorTypes = []
    if (!Validator.isMobileNumberCorrect(state.mobile)) {
      errorTypes.push({ errorType: 'wrongMobile' })
    }
    if (errorTypes.length) {
      reject({ type: types.SHOW_ERROR, payload: { errorTypes } })
    }
    allow(action)
  }
})

const requestJWT = createLogic({
  type: types.SUBMIT,
  process({ customerIamApiSdkJs, Storage, getState, action }, dispatch, done) {
    customerIamApiSdkJs
      .signIn(getJsState(getState).mobile, getJsState(getState).password)
      .then(res => {
        Storage.setJWT(res.data.token).then(() => {
          dispatch({
            type: types.SUBMIT_DONE
          })

          dispatch(tokenActions.setJWT(res.data.token))
          dispatch(NavigationActions.navigate({ routeName: 'Tab' }))
          done()
        })
      })
      .catch(error => {
        // setTimeout(() => {
        if (error === CustomerIamApiSdkJs.ERRORS.NETWORK_ERROR) {
          let errorTypes = []
          errorTypes.push({ errorType: 'networkError' })
          dispatch({
            type: types.SHOW_ERROR,
            payload: { errorTypes }
          })
          return done()
        }

        if (error === CustomerIamApiSdkJs.ERRORS.SERVER_ERROR) {
          let errorTypes = []
          errorTypes.push({ errorType: 'serverError' })
          dispatch({
            type: types.SHOW_ERROR,
            payload: { errorTypes }
          })
          return done()
        }

        if (error === CustomerIamApiSdkJs.ERRORS.USER_PASSWORD_DISMATCH) {
          let errorTypes = []
          errorTypes.push({ errorType: 'userPasswordMismatch' })
          dispatch({
            type: types.SHOW_ERROR,
            payload: { errorTypes }
          })
          return done()
        }

        let errorTypes = []
        errorTypes.push({ errorType: 'otherError' })
        dispatch({
          type: types.SHOW_ERROR,
          payload: { errorTypes }
        })
        done()
        // }, 3000)
      })
  }
})

export default [validate, requestJWT]
