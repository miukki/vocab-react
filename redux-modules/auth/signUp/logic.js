/* global setTimeout process */
/* global clearTimeout process */

import { createLogic } from 'redux-logic'
import CustomerIamApiSdkJs from 'customer-iam-api-sdk-js'
import * as tokenActions from '../token/actions'
import types from './types'
import Validator from '../../../lib/Validator'

const getJsState = function(getState) {
  return getState().toJS().signUp
}

const dispatchCountdownVcode = createLogic({
  type: [types.SEND_VCODE_DONE],
  process({ getState, action }, dispatch, done) {
    dispatch({ type: types.COUNTDOWN_VCODE })
    done()
  }
})

const validate = createLogic({
  type: [types.SUBMIT],
  validate({ getState, action }, allow, reject, done) {
    const state = getJsState(getState)
    let errorTypes = []
    if (!Validator.isVcodeCorrect(state.vcode)) {
      errorTypes.push({ errorType: 'wrongVcode' })
    }
    if (!Validator.isMobileNumberCorrect(state.mobile)) {
      errorTypes.push({ errorType: 'wrongMobile' })
    }
    if (!Validator.isStrongPassword(state.password)) {
      errorTypes.push({ errorType: 'weakPassword' })
    }
    if (errorTypes.length) {
      reject({ type: types.SHOW_ERROR, payload: { errorTypes } })
    }
    allow(action)
  }
})

const validateVCode = createLogic({
  type: [types.SEND_VCODE],
  validate({ getState, action }, allow, reject, done) {
    const state = getJsState(getState)
    let errorTypes = []
    if (!Validator.isMobileNumberCorrect(state.mobile)) {
      errorTypes.push({ errorType: 'wrongMobile' })
    }
    if (errorTypes.length) {
      reject({ type: types.SHOW_ERROR, payload: { errorTypes } })
    }
    allow(action)
  }
})

const sendVcode = createLogic({
  type: [types.SEND_VCODE],
  process({ customerIamApiSdkJs, getState, action }, dispatch, done) {
    customerIamApiSdkJs
      .sendSMS(getJsState(getState).mobile)
      .then(res => {
        dispatch({
          type: types.SEND_VCODE_DONE
        })
        done()
      })
      .catch(error => {
        // setTimeout(() => {
        if (error === CustomerIamApiSdkJs.ERRORS.WRONG_MOBILE_FORMAT) {
          let errorTypes = []
          errorTypes.push({ errorType: 'wrongMobile' })
          dispatch({
            type: types.SHOW_ERROR,
            payload: { errorTypes }
          })
          return done()
        }

        if (error === CustomerIamApiSdkJs.ERRORS.HIT_RATE_LIMIT) {
          let errorTypes = []
          errorTypes.push({ errorType: 'hitRateLimit' })
          dispatch({
            type: types.SHOW_ERROR,
            payload: { errorTypes }
          })
          return done()
        }

        if (error === CustomerIamApiSdkJs.ERRORS.NETWORK_ERROR) {
          let errorTypes = []
          errorTypes.push({ errorType: 'networkError' })
          dispatch({
            type: types.SHOW_ERROR,
            payload: { errorTypes }
          })
          return done()
        }

        if (error === CustomerIamApiSdkJs.ERRORS.SERVER_ERROR) {
          let errorTypes = []
          errorTypes.push({ errorType: 'serverError' })
          dispatch({
            type: types.SHOW_ERROR,
            payload: { errorTypes }
          })
          return done()
        }

        let errorTypes = []
        errorTypes.push({ errorType: 'otherError' })
        dispatch({
          type: types.SHOW_ERROR,
          payload: { errorTypes }
        })
        done()
        // }, 3000)
      })
  }
})

const signUp = createLogic({
  type: types.SUBMIT,
  process({ customerIamApiSdkJs, Storage, getState, action }, dispatch, done) {
    console.log('SUBMIT DATA TO SERVER')
    customerIamApiSdkJs
      // .resetPassword(getJsState(getState).mobile) // for test
      .signUp(
        getJsState(getState).vcode,
        getJsState(getState).mobile,
        getJsState(getState).password
      )
      .then(res => {
        Storage.setJWT(res.data.token).then(() => {
          dispatch(tokenActions.setJWT(res.data.token))
          dispatch({
            type: types.SUBMIT_DONE
          })
          done()
        })
      })
      .catch(error => {
        // setTimeout(() => {
        if (error === CustomerIamApiSdkJs.ERRORS.USER_ALREADY_EXIST) {
          let errorTypes = []
          errorTypes.push({ errorType: 'userAlreadyExist' })
          dispatch({
            type: types.SHOW_ERROR,
            payload: { errorTypes }
          })
          return done()
        }

        if (error === CustomerIamApiSdkJs.ERRORS.VERIFICATION_CODE_ERROR) {
          let errorTypes = []
          errorTypes.push({ errorType: 'wrongVcode' })
          dispatch({
            type: types.SHOW_ERROR,
            payload: { errorTypes }
          })
          return done()
        }

        if (error === CustomerIamApiSdkJs.ERRORS.NETWORK_ERROR) {
          let errorTypes = []
          errorTypes.push({ errorType: 'networkError' })
          dispatch({
            type: types.SHOW_ERROR,
            payload: { errorTypes }
          })
          return done()
        }

        if (error === CustomerIamApiSdkJs.ERRORS.SERVER_ERROR) {
          let errorTypes = []
          errorTypes.push({ errorType: 'serverError' })
          dispatch({
            type: types.SHOW_ERROR,
            payload: { errorTypes }
          })
          return done()
        }

        let errorTypes = []
        errorTypes.push({ errorType: 'otherError' })
        dispatch({
          type: types.SHOW_ERROR,
          payload: { errorTypes }
        })
        done()
        // }, 3000)
      })
  }
})

export default [validate, validateVCode, signUp, sendVcode]
