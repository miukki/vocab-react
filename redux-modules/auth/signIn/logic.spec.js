import { fromJS } from 'immutable'
import { combineReducers } from 'redux-immutable'
import { createMockStore } from 'redux-logic-test'
import CustomerIamApiSdkJs from 'customer-iam-api-sdk-js'

import reducer, { initialState } from './'
import types from './types'
import logic from './logic'

const rootReducer = combineReducers({ signIn: reducer })

describe('signIn', () => {
  describe('checkMobile', function() {
    it('reject action when mobile length wrong', function(done) {
      const store = createMockStore({
        initialState: fromJS({
          signIn: initialState.set('mobile', '1800000000')
        }),
        reducer: rootReducer,
        logic,
        injectedDeps: {}
      })

      store.dispatch({
        type: types.SUBMIT
      })

      store.whenComplete().then(() => {
        expect(store.actions).toMatchSnapshot()
        done()
      })
    })

    it('allow action when mobile OK', function(done) {
      const signIn = jest.fn()
      signIn.mockReturnValue(Promise.resolve({ data: { token: 'jwt-token' } }))

      const setJWT = jest.fn()
      setJWT.mockReturnValue(Promise.resolve())

      const store = createMockStore({
        initialState: fromJS({
          signIn: initialState.set('mobile', '18000000000')
        }),
        reducer: rootReducer,
        logic,
        injectedDeps: { customerIamApiSdkJs: { signIn }, Storage: { setJWT } }
      })

      store.dispatch({
        type: types.SUBMIT
      })

      store.whenComplete().then(() => {
        expect(store.actions).toMatchSnapshot()
        done()
      })
    })
  })

  describe('requestJWT', () => {
    it('request JWT from API and set it to storage', done => {
      const signIn = jest.fn()
      signIn.mockReturnValue(Promise.resolve({ data: { token: 'jwt-token' } }))

      const setJWT = jest.fn()
      setJWT.mockReturnValue(Promise.resolve())

      const store = createMockStore({
        initialState: fromJS({
          signIn: initialState
            .set('mobile', '18000000000')
            .set('password', 'Qweasdzxc123!!!')
        }),
        reducer: rootReducer,
        logic,
        injectedDeps: { customerIamApiSdkJs: { signIn }, Storage: { setJWT } }
      })

      store.dispatch({
        type: types.SUBMIT
      })

      store.whenComplete().then(() => {
        expect(signIn.mock.calls[0]).toMatchSnapshot()
        expect(setJWT.mock.calls[0]).toMatchSnapshot()
        expect(store.actions).toMatchSnapshot()
        done()
      })
    })

    it('dispatch SHOW_ERROR when networkError', done => {
      const signIn = jest.fn()
      signIn.mockReturnValue(
        Promise.reject(CustomerIamApiSdkJs.ERRORS.NETWORK_ERROR)
      )

      const store = createMockStore({
        initialState: fromJS({
          signIn: initialState
            .set('mobile', '18000000000')
            .set('password', 'Qweasdzxc123!!!')
        }),
        reducer: rootReducer,
        logic,
        injectedDeps: { customerIamApiSdkJs: { signIn } }
      })

      store.dispatch({
        type: types.SUBMIT
      })

      store.whenComplete().then(() => {
        expect(store.actions).toMatchSnapshot()
        done()
      })
    })

    it('dispatch SHOW_ERROR when serverError', done => {
      const signIn = jest.fn()
      signIn.mockReturnValue(
        Promise.reject(CustomerIamApiSdkJs.ERRORS.SERVER_ERROR)
      )

      const store = createMockStore({
        initialState: fromJS({
          signIn: initialState
            .set('mobile', '18000000000')
            .set('password', 'Qweasdzxc123!!!')
        }),
        reducer: rootReducer,
        logic,
        injectedDeps: { customerIamApiSdkJs: { signIn } }
      })

      store.dispatch({
        type: types.SUBMIT
      })

      store.whenComplete().then(() => {
        expect(store.actions).toMatchSnapshot()
        done()
      })
    })

    it('dispatch SHOW_ERROR when userPasswordMismatch', done => {
      const signIn = jest.fn()
      signIn.mockReturnValue(
        Promise.reject(CustomerIamApiSdkJs.ERRORS.USER_PASSWORD_DISMATCH)
      )

      const store = createMockStore({
        initialState: fromJS({
          signIn: initialState
            .set('mobile', '18000000000')
            .set('password', 'Qweasdzxc123!!!')
        }),
        reducer: rootReducer,
        logic,
        injectedDeps: { customerIamApiSdkJs: { signIn } }
      })

      store.dispatch({
        type: types.SUBMIT
      })

      store.whenComplete().then(() => {
        expect(store.actions).toMatchSnapshot()
        done()
      })
    })
  })
})
