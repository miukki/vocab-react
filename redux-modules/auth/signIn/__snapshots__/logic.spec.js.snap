// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`signIn checkMobile allow action when mobile OK 1`] = `
Array [
  Object {
    "type": "auth/signIn/SUBMIT",
  },
  Object {
    "type": "auth/signIn/SUBMIT_DONE",
  },
  Object {
    "payload": Object {
      "jwt": "jwt-token",
    },
    "type": "auth/token/SET_JWT",
  },
  Object {
    "routeName": "Tab",
    "type": "Navigation/NAVIGATE",
  },
]
`;

exports[`signIn checkMobile reject action when mobile length wrong 1`] = `
Array [
  Object {
    "payload": Object {
      "errorTypes": Array [
        Object {
          "errorType": "wrongMobile",
        },
      ],
    },
    "type": "auth/signIn/SHOW_ERROR",
  },
]
`;

exports[`signIn requestJWT dispatch SHOW_ERROR when networkError 1`] = `
Array [
  Object {
    "type": "auth/signIn/SUBMIT",
  },
  Object {
    "payload": Object {
      "errorTypes": Array [
        Object {
          "errorType": "networkError",
        },
      ],
    },
    "type": "auth/signIn/SHOW_ERROR",
  },
]
`;

exports[`signIn requestJWT dispatch SHOW_ERROR when serverError 1`] = `
Array [
  Object {
    "type": "auth/signIn/SUBMIT",
  },
  Object {
    "payload": Object {
      "errorTypes": Array [
        Object {
          "errorType": "serverError",
        },
      ],
    },
    "type": "auth/signIn/SHOW_ERROR",
  },
]
`;

exports[`signIn requestJWT dispatch SHOW_ERROR when userPasswordMismatch 1`] = `
Array [
  Object {
    "type": "auth/signIn/SUBMIT",
  },
  Object {
    "payload": Object {
      "errorTypes": Array [
        Object {
          "errorType": "userPasswordMismatch",
        },
      ],
    },
    "type": "auth/signIn/SHOW_ERROR",
  },
]
`;

exports[`signIn requestJWT request JWT from API and set it to storage 1`] = `
Array [
  "18000000000",
  "Qweasdzxc123!!!",
]
`;

exports[`signIn requestJWT request JWT from API and set it to storage 2`] = `
Array [
  "jwt-token",
]
`;

exports[`signIn requestJWT request JWT from API and set it to storage 3`] = `
Array [
  Object {
    "type": "auth/signIn/SUBMIT",
  },
  Object {
    "type": "auth/signIn/SUBMIT_DONE",
  },
  Object {
    "payload": Object {
      "jwt": "jwt-token",
    },
    "type": "auth/token/SET_JWT",
  },
  Object {
    "routeName": "Tab",
    "type": "Navigation/NAVIGATE",
  },
]
`;
